### ISSUE TITLE
bq_auth() on remote server using .json key -- won't authorize

### ISSUE THREAD

**Issue by reliscu** [r-dbi/bigrquery#408]
I run:
```
bq_auth(path="path/to/key.json")
```
Which triggers a browser window to open:

<img width="524" alt="Screen Shot 2020-10-13 at 16 01 09" src="https://user-images.githubusercontent.com/50351159/95924925-66aad480-0d6d-11eb-9899-98e5eb68a8cd.png">

I select the account associated with the key, then:

<img width="568" alt="Screen Shot 2020-10-13 at 16 05 02" src="https://user-images.githubusercontent.com/50351159/95925262-30218980-0d6e-11eb-8de5-96a8b621a31b.png">

I click allow, which brings me to this URL: 

http://localhost:1410/?state=...&code=4/...&scope=email%20openid%20https://www.googleapis.com/auth/userinfo.email%20https://www.googleapis.com/auth/cloud-platform%20https://www.googleapis.com/auth/bigquery&authuser=0&prompt=consent

...which greets me with "site can't be reached".

Is there a way to authorize in a non interactive manner / is the problem associated with running R on a remote server?

Session info:

```
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Devuan GNU/Linux 3 (beowulf)

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_1.0.2           preprocessCore_1.50.0 data.table_1.13.0     bigrquery_1.3.2      

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5       pillar_1.4.6     compiler_4.0.3   later_1.1.0.1    dbplyr_1.4.4     tools_4.0.3      bit_4.0.4       
 [8] jsonlite_1.7.1   lifecycle_0.2.0  tibble_3.0.3     gargle_0.5.0     pkgconfig_2.0.3  rlang_0.4.7      DBI_1.1.0       
[15] rstudioapi_0.11  curl_4.3         httr_1.4.2       generics_0.0.2   vctrs_0.3.4      fs_1.5.0         askpass_1.1     
[22] bit64_4.0.5      tidyselect_1.1.0 glue_1.4.2       R6_2.4.1         purrr_0.3.4      blob_1.2.1       magrittr_1.5    
[29] promises_1.1.1   ellipsis_0.3.1   assertthat_0.2.1 httpuv_1.5.4     crayon_1.3.4    
```

**Comment by jennybc** [comment]
What is in this `.json` file? How did you get it? If you paste here, replace anything that looks like a secret with placeholder text.

```
bq_auth(path="path/to/key.json")
```

Your syntax tells me you intend to provide:

```
path JSON identifying the service account
```

A very common mistake is to provide oauth app credentials (which can be downloaded as JSON) instead of a service account token (credentials which can also be downloaded as JSON).

Oauth credentials look like this (this would be the WRONG type of input to `path`):

```
{
    "installed": {
        "client_id": "STUFF",
        "project_id": "STUFF",
        "auth_uri": "https://accounts.google.com/o/oauth2/auth",
        "token_uri": "https://oauth2.googleapis.com/token",
        "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
        "client_secret": "STUFF",
        "redirect_uris": [
            "urn:ietf:wg:oauth:2.0:oob",
            "http://localhost"
        ]
    }
}
```

Service account token JSON looks like this (this would be the CORRECT type of input to `path`):

```
{
  "type": "service_account",
  "project_id": "STUFF",
  "private_key_id": "STUFF",
  "private_key": "-----BEGIN PRIVATE KEY-----\ STUFF\n-----END PRIVATE KEY-----\n",
  "client_email": "STUFF",
  "client_id": "STUFF",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "STUFF"
}
```

**Comment by reliscu** [comment]
Hi @jennybc,

My account token looks like your latter example:

```
{
  "type": "service_account",
  "project_id": "STUFF",
  "private_key_id": "STUFF"
  "private_key": "STUFF"
  "client_email": "STUFF",
  "client_id": "STUFF",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "STUFF"
}

````

**Comment by jennybc** [comment]
That's pretty odd then, because there should be no oauth dance.

If people do regular oauth flow on a server, they need to specify out-of-bound auth via `use_oob = TRUE`. But it sounds like you have and mean to use a service account token, so this does not apply to you.

Can you try to auth with this token locally? And/or you can turn on some extra debugging info via `options(gargle_quiet = FALSE)` as described here:

https://gargle.r-lib.org/articles/troubleshooting.html

I suspect the service account may not be set up correctly.

**Comment by reliscu** [comment]
Locally the key seems to work fine.

Having set `options(gargle_quiet = FALSE)`:

```
op <- options(gargle_quiet = FALSE)
gargle_quiet()

bq_auth(path="path/to/key.json")

trying token_fetch()
trying credentials_service_account()
adding 'userinfo.email' scope
Error: unable to load shared object '/home/shared/R/x86_64-pc-linux-gnu-library/4.0/openssl/libs/openssl.so':
  libssl.so.10: cannot open shared object file: No such file or directory
trying credentials_app_default()
trying credentials_gce()
trying credentials_byo_oauth()
Error: inherits(token, "Token2.0") is not TRUE
trying credentials_user_oauth2()
Gargle2.0 initialize
attempt from: bigrquery
adding 'userinfo.email' scope
loading token from the cache
no matching token in the cache
initiating new token
createTcpServer: address already in use
Error: Failed to create server
Error: Can't get Google credentials.
Are you running bigrquery in a non-interactive session? Consider:
  * Call `bq_auth()` directly with all necessary specifics.
```

**Comment by reliscu** [comment]
Update:

TLDR; openssl R package was corrupted.

---

openssl in R seemed to be corrupt, but would remain so after removing and reinstalling mutliple times. When a lab member installed it on their end, everything worked smoothly, INCLUDING  bq_auth() with the key listed above.

**Comment by jennybc** [comment]
> Locally the key seems to work fine.

That's not how I interpret what you show.

```
trying token_fetch()
trying credentials_service_account()
adding 'userinfo.email' scope
Error: unable to load shared object '/home/shared/R/x86_64-pc-linux-gnu-library/4.0/openssl/libs/openssl.so':
  libssl.so.10: cannot open shared object file: No such file or directory
```

So the service account token fails. I agree that your root problem is related to openssl and this is neither a bigrquery nor gargle problem.

**Comment by reliscu** [comment]
> > Locally the key seems to work fine.
> 
> That's not how I interpret what you show.
> 
> ```
> trying token_fetch()
> trying credentials_service_account()
> adding 'userinfo.email' scope
> Error: unable to load shared object '/home/shared/R/x86_64-pc-linux-gnu-library/4.0/openssl/libs/openssl.so':
>   libssl.so.10: cannot open shared object file: No such file or directory
> ```
> 
> So the service account token fails. I agree that your root problem is related to openssl and this is neither a bigrquery nor gargle problem.

I should have specified, the above output was again on the remote server. I didn't share any input/outputs from trying locally, which was successful.
